core/controllers/feedback_test.py

```diff
@@ -554,3 +556,216 @@ def test_feedback_threads_with_suggestions(self):
+            suggestion_thread_dicts = response_dict[
+                'suggestion_thread_dicts'][0]
+            topic_thread = feedback_services.get_all_threads(
+                feconf.ENTITY_TYPE_TOPIC, self.topic_id, True)[0]
+
+            self.assertEqual(suggestion_thread_dicts['subject'], 'a subject')
+            self.assertEqual(
+                suggestion_thread_dicts['thread_id'], topic_thread.id)
+        self.logout()
+
+
+class RecentFeedbackMessagesHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(RecentFeedbackMessagesHandlerTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.set_admins([self.OWNER_USERNAME])
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        self.exp_id = 'exp_id'
+
+    def test_get_recently_posted_feedback_messages(self):
+        self.login(self.OWNER_EMAIL)
+
+        response = self.get_json(
+            feconf.RECENT_FEEDBACK_MESSAGES_DATA_URL)
+
+        self.assertEqual(response['results'], [])
+        self.assertFalse(response['more'])
+
+        self.save_new_valid_exploration(
+            self.exp_id, self.owner_id, title='Exploration title',
+            category='Architecture', language_code='en')
+        feedback_services.create_thread(
+            feconf.ENTITY_TYPE_EXPLORATION, self.exp_id, self.owner_id,
+            'a subject', 'some text')
+
+        response = self.get_json(
+            feconf.RECENT_FEEDBACK_MESSAGES_DATA_URL)
+        results = response['results']
+
+        self.assertFalse(response['more'])
+        self.assertEqual(results[0]['text'], 'some text')
+        self.assertEqual(results[0]['updated_subject'], 'a subject')
+        self.assertEqual(results[0]['entity_type'], 'exploration')
+        self.assertEqual(results[0]['entity_id'], self.exp_id)
+
+        self.logout()
+
+
+class FeedbackStatsHandlerTests(test_utils.GenericTestBase):
+
+    def setUp(self):
+        super(FeedbackStatsHandlerTests, self).setUp()
+        self.signup(self.OWNER_EMAIL, self.OWNER_USERNAME)
+        self.set_admins([self.OWNER_USERNAME])
+        self.owner_id = self.get_user_id_from_email(self.OWNER_EMAIL)
+        self.exp_id = 'exp_id'
+
+    def test_get_num_threads_after_creating_feedback_analytics(self):
+        self.login(self.OWNER_EMAIL)
+
+        self.get_json(
+            '%s/%s' % (feconf.FEEDBACK_STATS_URL_PREFIX, self.exp_id),
+            expected_status_int=404)
+
+        self.save_new_valid_exploration(
+            self.exp_id, self.owner_id, title='Exploration title',
+            category='Architecture', language_code='en')
+
+        response = self.get_json(
+            '%s/%s' % (feconf.FEEDBACK_STATS_URL_PREFIX, self.exp_id))
+
+        self.assertEqual(response['num_total_threads'], 0)
+        self.assertEqual(response['num_open_threads'], 0)
+
+        feedback_models.FeedbackAnalyticsModel.create(self.exp_id, 2, 3)
```

Comment:
I don't think you should manipulate and create models in the datastore directly. You should be able to generate these models by running some services (or continuous computation jobs) after creating threads. Use that to generate the data you are querying.

Reply:
OK I ran a continuous computation job to generate the data in this test class. But I did not found one for creating suggestions.

Reply:
Oh, I was talking only about the analytics model. Creating them directly was kinda hacky and I didn't want you to do that...

Reply:
Hmm agreed, we should not test hard-coded data when we have a service perfectly capable of doing it for us. @seanlip WDYT?

Reply:
Yes, I agree with not directly creating models in the datastore unless there is actually a strong reason for doing so. Use the existing functions in the domain/services layer for that.